fix: get `supportedChains` to avoid blocking the confirmation process

vinistevam opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR addresses the blocking behaviour in the confirmation process caused by the synchronous `supportedChains` check. Previously, the Send flow was halted until a response from the `security-alerts.api.cx.metamask.io` API was received. This created delays and negatively impacted the user experience, especially when the API response was slow.

**Key Changes:**
**Asynchronous Chain Validation:** The `supportedChains` check is now performed asynchronously within the non-awaited `validateRequestWithPPOM` function. This ensures that the Send flow is not blocked while waiting for the API response.
**Introduction of CheckingChain Result Type:** A new result type, `CheckingChain`, is introduced to represent the intermediate state before a definitive result is obtained. This state is used before the `loading` status in both middleware and transaction utility functions.

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28313?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/28255 https://github.com/MetaMask/metamask-extension/issues/28257

## **Manual testing steps**

1. Use a proxy to intercept requests
2. Start a Send flow
3. Intercept the request to `security-alerts.api.cx.metamask.io`
4. See send flow cannot be initiated until we have a response to this request

## **Screenshots/Recordings**

[supported-chains-.webm](https://github.com/user-attachments/assets/4e9e495a-10f3-4bb1-8d05-8045a735b655)

[unsupported-chains.webm](https://github.com/user-attachments/assets/e5767bc1-2eab-44bd-83c3-777d34c23ff6)

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

vinistevam committed:
fix: get supportedChains async to avoid blocking the confirmation process

vinistevam committed:
Merge branch 'develop' into fix/supported-chains-blocks-sendflow

vinistevam committed:
fix e2e

vinistevam committed:
fix e2e

vinistevam committed:
fix e2e

vinistevam committed:
Merge branch 'develop' into fix/supported-chains-blocks-sendflow

vinistevam committed:
fix e2e

vinistevam committed:
Merge branch 'develop' into fix/supported-chains-blocks-sendflow

matthewwalsh0 commented on app/scripts/lib/ppom/ppom-util.ts:
```diff
@@ -37,18 +39,37 @@ type PPOMRequest = Omit<JsonRpcRequest, 'method' | 'params'> & {
 }): Promise<SecurityAlertResponse> {
   try {
+    if (!(await isChainSupported(chainId))) {
+      return {
```
We're returning this new response but should we also update so the controllers have the correct chain not supported response on them?

vinistevam replied:
That is true. I pushed a change to update it before responding.

matthewwalsh0 commented on app/scripts/lib/ppom/ppom-util.ts:
```diff
@@ -37,18 +39,37 @@ type PPOMRequest = Omit<JsonRpcRequest, 'method' | 'params'> & {
+      };
+    }
+
+    updateSecurityResponse(request.method, securityAlertId, {
```
Isn't this asynchronous, should we await this before continuing?

vinistevam replied:
You are right, the type previous was void but because we were intentionally not awaiting. Fixed.

matthewwalsh0 commented on shared/constants/security-provider.ts:
```diff
@@ -32,7 +32,7 @@ export enum BlockaidReason {
   /** Malicious signature on Blur order  */
   blurFarming = 'blur_farming',
-  /** A known malicous site invoked that transaction  */
+  /** A known malicious site invoked that transaction  */
```
I'm certain you have a spell checker VSCode extension to catch all these 😄

matthewwalsh0 commented on shared/constants/security-provider.ts:
```diff
@@ -117,6 +119,17 @@ export const LOADING_SECURITY_ALERT_RESPONSE: SecurityAlertResponse = {
+  {
+    result_type: BlockaidResultType.Benign,
+    reason: BlockaidReason.chainNotSupported,
+  };
```
Minor, extra indent here?

matthewwalsh0 commented on test/e2e/tests/metrics/signature-approved.spec.js:
```diff
@@ -47,6 +52,16 @@ async function mockSegment(mockServer) {
   ];
 }
 
+const eventPropertiesBase = {
```
Minor, `expectedEventPropertiesBase`?

vinistevam replied:
Done

vinistevam committed:
apply suggestions

vinistevam committed:
Merge branch 'develop' into fix/supported-chains-blocks-sendflow

vinistevam committed:
fix e2e

vinistevam committed:
applied suggestions

vinistevam commented on app/scripts/lib/transaction/util.test.ts:
```diff
@@ -425,8 +425,8 @@ describe('Transaction Utils', () => {
-            reason: BlockaidReason.inProgress,
-            result_type: BlockaidResultType.Loading,
+            reason: BlockaidReason.checkingChain,
+            result_type: BlockaidResultType.Benign,
```
It was changed to `loading`.

matthewwalsh0 commented on app/scripts/lib/ppom/ppom-middleware.ts:
```diff
@@ -122,6 +115,7 @@ export function createPPOMMiddleware<
             chainId,
+            updateSecurityAlertResponse,
           }).then((securityAlertResponse) => {
             updateSecurityAlertResponse(
```
As we're passing the update callback to the util, can we omit this `then` entirely plus the one in `transaction/util` since the util can own all the updates?

vinistevam committed:
applied suggestions

vinistevam committed:
Merge branch 'develop' into fix/supported-chains-blocks-sendflow

vinistevam committed:
fix: use `updateSecurityResponse` (#28418)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28418?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.
